File: src/collectors/aggregate.ts
interface ContributionDay {
  date: string;
  contributionCount: number;
  color: string;
}

interface CalendarWeek {
  contributionDays: ContributionDay[];
}

function extractProfileSummary(results: Record<string, unknown>, username: string): Record<string, unknown> {
  const profile = (results.profile as Record<string, unknown>) ?? {};
  const contributions = (results.contributions as Record<string, unknown>) ?? {};
  const repos = (results.repositories as Record<string, unknown>[]) ?? [];
  return {
    username,
    name: profile.name || username,
    avatarUrl: profile.avatarUrl || `https://avatars.githubusercontent.com/u/${username}`,
    bio: profile.bio ?? null,
    company: profile.company ?? null,
    location: profile.location ?? null,
    followers: profile.followers ?? 0,
    following: profile.following ?? 0,
    contributions: contributions.totalContributions ?? 0,
    currentStreak: contributions.currentStreak ?? 0,
    repos: repos.length,
    stars: repos.reduce((a, r) => a + ((r.stars as number) ?? 0), 0),
  };
}

function parseTimestamp(iso: string): number {
  try {
    return new Date(iso).getTime();
  } catch {
    return 0;
  }
}

function aggregateContributions(contributionsList: Record<string, unknown>[]): Record<string, unknown> {
  if (contributionsList.length === 0) return {};

  let total = 0;
  let totalCommit = 0;
  let totalIssue = 0;
  let totalPR = 0;
  let totalReview = 0;
  let restricted = 0;
  let currentStreak = 0;
  let longestStreak = 0;
  const calendarWeeks: CalendarWeek[] = [];

  for (const contrib of contributionsList) {
    total += (contrib.totalContributions as number) ?? 0;
    totalCommit += (contrib.totalCommitContributions as number) ?? 0;
    totalIssue += (contrib.totalIssueContributions as number) ?? 0;
    totalPR += (contrib.totalPullRequestContributions as number) ?? 0;
    totalReview += (contrib.totalPullRequestReviewContributions as number) ?? 0;
    restricted += (contrib.restrictedContributionsCount as number) ?? 0;
    currentStreak = Math.max(currentStreak, (contrib.currentStreak as number) ?? 0);
    longestStreak = Math.max(longestStreak, (contrib.longestStreak as number) ?? 0);
    const weeks = (contrib.calendar as CalendarWeek[]) ?? [];
    calendarWeeks.push(...weeks);
  }

  const mergedCalendar = mergeCalendars(calendarWeeks);

  return {
    totalContributions: total,
    totalCommitContributions: totalCommit,
    totalIssueContributions: totalIssue,
    totalPullRequestContributions: totalPR,
    totalPullRequestReviewContributions: totalReview,
    restrictedContributionsCount: restricted,
    currentStreak,
    longestStreak,
    calendar: mergedCalendar,
  };
}

const COLOR_SCALE = ["#ebedf0", "#9be9a8", "#40c463", "#30a14e", "#216e39"];

function upgradeColor(current: string, newColor: string): string {
  const currentIdx = COLOR_SCALE.indexOf(current);
  const newIdx = COLOR_SCALE.indexOf(newColor);
  if (newIdx > currentIdx) return newColor;
  return current;
}

function mergeCalendars(weeks: CalendarWeek[]): CalendarWeek[] {
  const dayCounts: Record<string, ContributionDay> = {};
  for (const week of weeks) {
    for (const day of week.contributionDays) {
      const date = day.date;
      if (!date) continue;
      if (!dayCounts[date]) {
        dayCounts[date] = { date, contributionCount: day.contributionCount ?? 0, color: day.color ?? "#ebedf0" };
      } else {
        dayCounts[date].contributionCount += day.contributionCount ?? 0;
        dayCounts[date].color = upgradeColor(dayCounts[date].color, day.color ?? "#ebedf0");
      }
    }
  }

  const sortedDays = Object.values(dayCounts).sort((a, b) => a.date.localeCompare(b.date));
  const mergedWeeks: CalendarWeek[] = [];
  let currentWeek: ContributionDay[] = [];
  for (const day of sortedDays) {
    currentWeek.push(day);
    if (currentWeek.length === 7) {
      mergedWeeks.push({ contributionDays: currentWeek });
      currentWeek = [];
    }
  }
  if (currentWeek.length > 0) {
    mergedWeeks.push({ contributionDays: currentWeek });
  }

  return mergedWeeks;
}

function aggregateActivities(activities: Record<string, unknown>[]): Record<string, number> {
  const result: Record<string, number> = {};
  for (const activity of activities) {
    for (const [key, value] of Object.entries(activity)) {
      if (typeof value === "number") {
        result[key] = (result[key] ?? 0) + value;
      }
    }
  }
  return result;
}

function aggregateDiscussions(discussionsList: Record<string, unknown>[]): Record<string, number> {
  const result: Record<string, number> = {};
  for (const discussions of discussionsList) {
    for (const [key, value] of Object.entries(discussions)) {
      if (typeof value === "number") {
        result[key] = (result[key] ?? 0) + value;
      }
    }
  }
  return result;
}

export function aggregateResults(
  resultsByUser: Record<string, Record<string, unknown>>
): Record<string, unknown> {
  if (Object.keys(resultsByUser).length === 0) return {};

  if (Object.keys(resultsByUser).length === 1) {
    const [username, results] = Object.entries(resultsByUser)[0];
    return { ...results, profiles: [extractProfileSummary(results, username)] };
  }

  const profiles: Record<string, unknown>[] = [];
  const allRepos: Record<string, unknown>[] = [];
  const allContributions: Record<string, unknown>[] = [];
  const allGists: Record<string, unknown>[] = [];
  const allTrafficViews: Record<string, unknown>[] = [];
  const allTrafficClones: Record<string, unknown>[] = [];
  const allContributorStats: Record<string, unknown>[] = [];
  const allActivity: Record<string, unknown>[] = [];
  const allDiscussions: Record<string, unknown>[] = [];
  const allStarsGiven: Record<string, unknown>[] = [];
  const allRepoContributions: Record<string, Record<string, number>> = {};
  let primaryProfile: Record<string, unknown> = {};

  for (const [username, results] of Object.entries(resultsByUser)) {
    profiles.push(extractProfileSummary(results, username));

    if (Object.keys(primaryProfile).length === 0 && results.profile) {
      primaryProfile = { ...(results.profile as Record<string, unknown>), username };
    }

    if (results.repositories) {
      allRepos.push(...(results.repositories as Record<string, unknown>[]));
    }
    if (results.contributions) {
      allContributions.push(results.contributions as Record<string, unknown>);
    }
    if (results.gists) {
      allGists.push(...(results.gists as Record<string, unknown>[]));
    }
    if (results.traffic) {
      const traffic = results.traffic as Record<string, unknown>;
      const views = traffic.views as Array<Record<string, unknown>> | undefined;
      const clones = traffic.clones as Array<Record<string, unknown>> | undefined;
      if (views) allTrafficViews.push(...views);
      if (clones) allTrafficClones.push(...clones);
    }
    if (results.contributor_stats) {
      allContributorStats.push(...(results.contributor_stats as Record<string, unknown>[]));
    }
    if (results.activity) {
      allActivity.push(results.activity as Record<string, unknown>);
    }
    if (results.discussions) {
      allDiscussions.push(results.discussions as Record<string, unknown>);
    }
    if (results.stars_given) {
      allStarsGiven.push(...(results.stars_given as Record<string, unknown>[]));
    }
    if (results.repo_contributions) {
      const repoContrib = results.repo_contributions as Record<string, Record<string, number>>;
      for (const [fullName, counts] of Object.entries(repoContrib)) {
        if (!allRepoContributions[fullName]) {
          allRepoContributions[fullName] = { ...counts };
        } else {
          for (const [key, value] of Object.entries(counts)) {
            allRepoContributions[fullName][key] = (allRepoContributions[fullName][key] ?? 0) + value;
          }
        }
      }
    }
  }

  // Deduplicate repos by full_name (keep most recent push)
  const seenRepos = new Map<string, Record<string, unknown>>();
  for (const repo of allRepos) {
    const fullName = repo.full_name as string;
    if (!fullName) continue;
    const existing = seenRepos.get(fullName);
    if (!existing || parseTimestamp(repo.pushed_at as string) > parseTimestamp(existing.pushed_at as string)) {
      seenRepos.set(fullName, repo);
    }
  }
  const dedupedRepos = Array.from(seenRepos.values());

  const aggregatedContributions = aggregateContributions(allContributions);
  const aggregatedActivity = aggregateActivities(allActivity);
  const aggregatedDiscussions = aggregateDiscussions(allDiscussions);

  return {
    profile: primaryProfile,
    profiles,
    contributions: aggregatedContributions,
    repositories: dedupedRepos,
    gists: allGists,
    traffic: { views: allTrafficViews, clones: allTrafficClones },
    contributor_stats: allContributorStats,
    activity: aggregatedActivity,
    discussions: aggregatedDiscussions,
    stars_given: allStarsGiven,
    repo_contributions: allRepoContributions,
  };
}

File: src/diffler/collectors/aggregate.py
"""Aggregate data from multiple GitHub profiles into combined stats.

When a user configures multiple usernames (e.g. personal + work),
diffler collects data for each profile independently and then merges
the results so templates see both individual profiles and combined
totals.
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


def aggregate_results(results_by_user: dict[str, dict[str, Any]]) -> dict[str, Any]:
    """Merge per-user collector results into combined stats.

    Args:
        results_by_user: Dict mapping username -> collector results dict.

    Returns:
        Aggregated results dict with the same shape as a single-user
        result, plus a ``profiles`` list for template access.
    """
    if not results_by_user:
        return {}

    if len(results_by_user) == 1:
        # Single user — no aggregation needed
        _username, results = next(iter(results_by_user.items()))
        results["profiles"] = [_extract_profile_summary(results, _username)]
        return results

    profiles: list[dict[str, Any]] = []
    all_repos: list[dict] = []
    all_contributions: list[dict] = []
    all_gists: list[dict] = []
    all_traffic_views: list[dict] = []
    all_traffic_clones: list[dict] = []
    all_contributor_stats: list[dict] = []
    all_activity: list[dict] = []
    all_discussions: list[dict] = []
    all_stars_given: list[dict] = []
    all_repo_contributions: dict[str, dict[str, int]] = {}
    primary_profile: dict[str, Any] = {}

    for username, results in results_by_user.items():
        profiles.append(_extract_profile_summary(results, username))

        if not primary_profile and results.get("profile"):
            primary_profile = dict(results["profile"])
            primary_profile["username"] = username

        if repos := results.get("repositories"):
            all_repos.extend(repos)

        if contrib := results.get("contributions"):
            all_contributions.append(contrib)

        if gists := results.get("gists"):
            all_gists.extend(gists)

        if traffic := results.get("traffic", {}):
            if views := traffic.get("views"):
                all_traffic_views.extend(views)
            if clones := traffic.get("clones"):
                all_traffic_clones.extend(clones)

        if stats := results.get("contributor_stats"):
            all_contributor_stats.extend(stats)

        if activity := results.get("activity"):
            all_activity.append(activity)

        if discussions := results.get("discussions"):
            all_discussions.append(discussions)

        if stars := results.get("stars_given"):
            all_stars_given.extend(stars)

        if repo_contrib := results.get("repo_contributions"):
            for full_name, counts in repo_contrib.items():
                if full_name not in all_repo_contributions:
                    all_repo_contributions[full_name] = dict(counts)
                else:
                    for key, value in counts.items():
                        all_repo_contributions[full_name][key] = (
                            all_repo_contributions[full_name].get(key, 0) + value
                        )

    # Deduplicate repos by full_name (keep the one with most recent push)
    seen_repos: dict[str, dict] = {}
    for repo in all_repos:
        full_name = repo.get("full_name", "")
        if not full_name:
            continue
        if full_name not in seen_repos:
            seen_repos[full_name] = repo
        else:
            # Keep the repo with the most recent push_at
            existing = seen_repos[full_name]
            if _parse_ts(repo.get("pushed_at", "")) > _parse_ts(existing.get("pushed_at", "")):
                seen_repos[full_name] = repo
    deduped_repos = list(seen_repos.values())

    # Aggregate contributions
    aggregated_contributions = _aggregate_contributions(all_contributions)

    # Aggregate activity
    aggregated_activity = _aggregate_activities(all_activity)

    # Aggregate discussions
    aggregated_discussions = _aggregate_discussions(all_discussions)

    return {
        "profile": primary_profile,
        "profiles": profiles,
        "contributions": aggregated_contributions,
        "repositories": deduped_repos,
        "gists": all_gists,
        "traffic": {
            "views": all_traffic_views,
            "clones": all_traffic_clones,
        },
        "contributor_stats": all_contributor_stats,
        "activity": aggregated_activity,
        "discussions": aggregated_discussions,
        "stars_given": all_stars_given,
        "repo_contributions": all_repo_contributions,
    }


def _extract_profile_summary(results: dict[str, Any], username: str) -> dict[str, Any]:
    """Extract a lightweight summary for the ``profiles`` list."""
    profile = results.get("profile", {})
    contributions = results.get("contributions", {})
    repos = results.get("repositories", [])
    return {
        "username": username,
        "name": profile.get("name") or username,
        "avatarUrl": profile.get("avatarUrl") or f"https://avatars.githubusercontent.com/u/{username}",
        "bio": profile.get("bio"),
        "company": profile.get("company"),
        "location": profile.get("location"),
        "followers": profile.get("followers", 0),
        "following": profile.get("following", 0),
        "contributions": contributions.get("totalContributions", 0),
        "currentStreak": contributions.get("currentStreak", 0),
        "repos": len(repos),
        "stars": sum(r.get("stars", 0) for r in repos),
    }


def _aggregate_contributions(contributions_list: list[dict[str, Any]]) -> dict[str, Any]:
    """Sum contributions across multiple profiles."""
    if not contributions_list:
        return {}

    total = 0
    total_commit = 0
    total_issue = 0
    total_pr = 0
    total_review = 0
    restricted = 0
    current_streak = 0
    longest_streak = 0
    calendar_weeks: list[dict] = []

    for contrib in contributions_list:
        total += contrib.get("totalContributions", 0)
        total_commit += contrib.get("totalCommitContributions", 0)
        total_issue += contrib.get("totalIssueContributions", 0)
        total_pr += contrib.get("totalPullRequestContributions", 0)
        total_review += contrib.get("totalPullRequestReviewContributions", 0)
        restricted += contrib.get("restrictedContributionsCount", 0)
        current_streak = max(current_streak, contrib.get("currentStreak", 0))
        longest_streak = max(longest_streak, contrib.get("longestStreak", 0))
        if weeks := contrib.get("calendar", []):
            calendar_weeks.extend(weeks)

    # Merge calendar weeks by summing contribution counts per day
    merged_calendar = _merge_calendars(calendar_weeks)

    return {
        "totalContributions": total,
        "totalCommitContributions": total_commit,
        "totalIssueContributions": total_issue,
        "totalPullRequestContributions": total_pr,
        "totalPullRequestReviewContributions": total_review,
        "restrictedContributionsCount": restricted,
        "currentStreak": current_streak,
        "longestStreak": longest_streak,
        "calendar": merged_calendar,
    }


def _merge_calendars(weeks: list[dict]) -> list[dict]:
    """Merge contribution calendar weeks, summing counts per day."""
    day_counts: dict[str, dict] = {}
    for week in weeks:
        for day in week.get("contributionDays", []):
            date = day.get("date", "")
            if not date:
                continue
            if date not in day_counts:
                day_counts[date] = {
                    "date": date,
                    "contributionCount": day.get("contributionCount", 0),
                    "color": day.get("color", "#ebedf0"),
                }
            else:
                day_counts[date]["contributionCount"] += day.get("contributionCount", 0)
                # Upgrade color if activity increased
                day_counts[date]["color"] = _upgrade_color(
                    day_counts[date]["color"], day.get("color", "#ebedf0")
                )

    # Re-group into weeks (7-day chunks sorted by date)
    sorted_days = sorted(day_counts.values(), key=lambda d: d["date"])
    merged_weeks: list[dict] = []
    current_week: list[dict] = []
    for day in sorted_days:
        current_week.append(day)
        if len(current_week) == 7:
            merged_weeks.append({"contributionDays": current_week})
            current_week = []
    if current_week:
        merged_weeks.append({"contributionDays": current_week})

    return merged_weeks


def _upgrade_color(current: str, new: str) -> str:
    """Return the more active color."""
    # GitHub color scale: #ebedf0 < #9be9a8 < #40c463 < #30a14e < #216e39
    scale = ["#ebedf0", "#9be9a8", "#40c463", "#30a14e", "#216e39"]
    try:
        return max(current, new, key=lambda c: scale.index(c) if c in scale else -1)
    except ValueError:
        return current


def _aggregate_activities(activities: list[dict[str, Any]]) -> dict[str, Any]:
    """Sum activity counts across profiles."""
    result: dict[str, int] = {}
    for activity in activities:
        for key, value in activity.items():
            if isinstance(value, int):
                result[key] = result.get(key, 0) + value
    return result


def _aggregate_discussions(discussions_list: list[dict[str, Any]]) -> dict[str, Any]:
    """Sum discussion counts across profiles."""
    result: dict[str, int] = {}
    for discussions in discussions_list:
        for key, value in discussions.items():
            if isinstance(value, int):
                result[key] = result.get(key, 0) + value
    return result


def _parse_ts(iso: str) -> float:
    """Parse ISO timestamp to Unix seconds."""
    from datetime import datetime
    try:
        dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
        return dt.timestamp()
    except Exception:
        return 0.0
